crates/mdb/src/lib.rs

```diff
@@ -218,80 +160,75 @@ impl Drop for SubscriberConfig {
         // and no other code reads it, so it is safe to drop.
         unsafe {
             mdb_sys::mdb_subscriber_config_destroy(&mut self.ptr);
-            if !self.on_message.is_null() {
-                drop(Box::from_raw(self.on_message));
-            }
+            drop(Box::from_raw(self.on_message))
         }
     }
 }
 
 pub struct Subscriber<'a> {
-    // Ensure the raw connection is not destroyed before the subscriber
-    _connection: &'a Connection,
     ptr: *mut mdb_sys::mdb_subscriber_t,
-    on_done: *mut Box<OnDone>,
-    // We don't need to keep the entire config alive, only the callback, because
-    // `mdb_subscriber_create_async` will copy any information it keeps.
-    on_message: *mut Box<OnMessage>,
+    on_done: *mut OnDone,
+    _marker: PhantomData<&'a Connection>,
 }
 
 impl<'a> Subscriber<'a> {
-    pub fn try_new(
+    pub fn try_new<F>(
         connection: &'a Connection,
-        mut config: SubscriberConfig,
-        on_done: Box<OnDone>,
-    ) -> Result<Self, Error> {
+        config: SubscriberConfig,
+        on_done: F,
+    ) -> Result<Self, OwnedError>
+    where
+        F: for<'b> Callback<Option<BorrowedError<'b>>>,
+    {
         debug!("Creating {}...", any::type_name::<Self>());
-        unsafe {
-            let on_done = Box::into_raw(Box::new(on_done));
-            let mut error: *mut mdb_sys::mdb_error_t = std::ptr::null_mut();
-            let ptr = mdb_sys::mdb_subscriber_create_async(
+        let on_done = Box::into_raw(Box::new(on_done));
+        let mut error: *mut mdb_sys::mdb_error_t = std::ptr::null_mut();
+        let ptr = unsafe {
+            mdb_sys::mdb_subscriber_create_async(
                 connection.ptr,
                 config.ptr,
-                Some(Self::on_done),
+                Some(Self::on_done::<F>),
                 on_done as *mut c_void,
                 &mut error,
-            );
-            match (ptr.is_null(), error.is_null()) {
-                (false, false) => {
-                    panic!("mdb_subscriber_create_async returned both a connection and an error")
-                }
-                (false, true) => {
-                    let on_message = config.on_message;
-                    config.on_message = std::ptr::null_mut();
```

Comment:
If we don't store the `on_message` callback on the `Subscriber`, will it not be dropped as soon as this `try_new` returns?

Reply:
Good point. Could Subscriber just own SubscriberConfig which then is dropped when Subscriber is dropped?

Reply:
It could. I actually wrote it like that at first to minimize the complexity, but then I changed my mind thinking that it is unnecessary to tie up the `mdb_subscriber_config_t` for the lifetime of the connection. Are there reasons other than complexity for going back? Is there anything in particular that you think contributes to complexity with the current solution (assuming complexity is the reason you want to change this)?